autoload/ale/completion.vim

```diff
@@ -106,7 +106,7 @@ function! ale#completion#Filter(buffer, filetype, suggestions, prefix) abort
     "   foo.
     "       ^
     " We need to include all of the given suggestions.
-    if index(l:triggers, a:prefix) >= 0
+    if index(l:triggers, a:prefix) >= 0 || empty(a:prefix)
```

Comment:
Instead of writing this here, move `if empty(a:prefix)` to the top of the function, before `l:triggers` is loaded, and return `a:suggestions` right away. That will be more efficient when there's no prefix.

Reply:
In `GetTriggerCharacter` above, you should probably add `if empty(a:prefix) ... return ''` to the top of the function, to avoid processing we don't need again.

Reply:
Ah, I said _return_ `a:suggestions` right away. What I should say is set the variable right away without the other processing above. The suggestions still need to be filtered below, as some LSP configurations have filters in Vim for removing noise.

Reply:
Add a test which ensures that Python filtering on suggestions still works when there's no prefix.

Reply:
Roger! When you say:

> The suggestions still need to be filtered below, as some LSP configurations have filters in Vim for removing noise.

are you referring to the block starting with `if !empty(l:excluded_words)`?

Reply:
> Add a test which ensures that filtering on suggestions for Python still works when there's no prefix.

Do you think the test I added to `test_completion_filtering.vader` is enough?

```vim
  AssertEqual
  \ ['Deutsch'],
  \ ale#completion#Filter(bufnr(''), '', ['Deutsch'], '')
```

Reply:
Yes, we the function shouldn't return early, I realised, and should continue down to the part for processing filtering with functions. Look through the tests and make sure that we have a test for the Python language server filtering function which still filters the results when the prefix is empty.